rust-src/concordium_base/src/curve_arithmetic/arkworks_instances.rs

```diff
@@ -0,0 +1,207 @@
+/// implements `ark_ec::CurveGroup`, provided an instance of `ArkCurveConfig`
+/// for that curve.
+impl<G: ark_ec::CurveGroup + ArkCurveConfig<G>> Curve for ArkGroup<G>
+where
+    <G as ark_ec::Group>::ScalarField: Serialize,
+{
+    type MultiExpType = GenericMultiExp<Self>;
+    type Scalar = ArkField<<G as ark_ec::Group>::ScalarField>;
+
+    const GROUP_ELEMENT_LENGTH: usize = G::GROUP_ELEMENT_LENGTH;
+    const SCALAR_LENGTH: usize = G::SCALAR_LENGTH;
+
+    fn zero_point() -> Self { ArkGroup(G::zero()) }
+
+    fn one_point() -> Self { ArkGroup(G::generator()) }
+
+    fn is_zero_point(&self) -> bool { self.0.is_zero() }
+
+    fn inverse_point(&self) -> Self { ArkGroup(-self.0) }
+
+    fn double_point(&self) -> Self { ArkGroup(self.0.double()) }
+
+    fn plus_point(&self, other: &Self) -> Self { ArkGroup(self.0 + other.0) }
+
+    fn minus_point(&self, other: &Self) -> Self { ArkGroup(self.0 - other.0) }
+
+    fn mul_by_scalar(&self, scalar: &Self::Scalar) -> Self { ArkGroup(self.0 * scalar.0) }
+
+    fn bytes_to_curve_unchecked<R: byteorder::ReadBytesExt>(b: &mut R) -> anyhow::Result<Self> {
+        // TODO: this is not the most efficient implementation, since there might be
+        // some additional checks during deserialization. However, it seems
+        // there are no unchecked methods available through traits.
+        let res = G::Affine::deserialize_compressed(b).map_err(|e| anyhow!(e))?;
+        Ok(ArkGroup(res.into()))
+    }
+
+    fn generate<R: rand::prelude::Rng>(rng: &mut R) -> Self { ArkGroup(G::rand(rng)) }
+
+    fn generate_scalar<R: rand::prelude::Rng>(rng: &mut R) -> Self::Scalar {
+        <G::ScalarField as ark_ff::UniformRand>::rand(rng).into()
+    }
+
+    fn scalar_from_u64(n: u64) -> Self::Scalar { ArkField(G::ScalarField::from(n)) }
+
+    fn scalar_from_bytes<A: AsRef<[u8]>>(bs: A) -> Self::Scalar {
+        // Traverse at most 4 8-byte chunks, for a total of 256 bits.
+        // The top-most two bits in the last chunk are set to 0.
+        let s = num::integer::div_ceil(Self::Scalar::CAPACITY, 8);
+        let mut fr = Vec::with_capacity(s as usize);
+        for chunk in bs.as_ref().chunks(8).take(s as usize) {
+            let mut v = [0u8; 8];
+            v[..chunk.len()].copy_from_slice(chunk);
+            fr.push(u64::from_le_bytes(v));
+        }
+        // unset two topmost bits in the last read u64.
+        *fr.last_mut().expect("Non empty vector expected") &= !(1u64 << 63 | 1u64 << 62);
+        // fr[3] &= !(1u64 << 63 | 1u64 << 62);
+        <Self::Scalar>::from_repr(&fr)
+            .expect("The scalar with top two bits erased should be valid.")
+    }
+
+    fn hash_to_group(m: &[u8]) -> Self {
+        let hasher = G::Hasher::new(G::DOMAIN_STRING.as_ref())
+            .expect("Expected valid domain separation string");
+        let res = G::Hasher::hash(&hasher, m).expect("Expected successful hashing to curve");
```

Comment:
It's unclear to me in which cases this fails?

Can it happen? 

It's not ideal to have all of these unwraps here since it makes the implemenation not generic.

Would it complicate a lot to propagate the failure?

Reply:
It is changed now to return `Result`. It the usage points, the return values are unwrapped.